[DM-28897] Initial version

rra opened the pull request.

rra committed:
Initial version

rra committed:
Add read/write ACL semantics as a variation

rra committed:
Updates from feedback

Explicitly discuss public collections and collections that are not
associated with either a user or a group.  Add a public flag to the
ACL and short-circuit access to public collections before making a
call to the identity management system.

Add discussion of using GIDs instead of groups and future work to
make that change.

Be clearer about who can modify the ACL.

rra committed:
Remove the public flag

The public flag complicates the ACL logic and probably isn't
necessary, so move that to variants.  Add some commentary about
non-user, non-group collections with ACLs.

ktlim commented on index.rst:
```diff
@@ -1,61 +1,190 @@
+========
 
-  Use the following syntax for sections:
+Given the authentication implementation described in DMTN-169_, Butler then has to decide whether to permit an operation on a collection following requirement DMS-REQ-0340 in LSE-61_.
```
DMTN-169 says that access to the database will go through a server that performs authorization checks and that pre-signed URLs will be returned to the user to allow authorized access to datasets.  But neither that document nor this one presents an argument that solely determining authorization by collection is sufficient to meet requirement 3 in [DMTN-169 §3.1](https://dmtn-169.lsst.io/#requirements).

Among other things, such an argument would seem to have to say that all operations require a collection and that no datasets can be added to a collection (directly or via collection chaining) unless the user has authorization to access them.

rra replied:
I added some language saying that all Butler operations are done via a collection.

ktlim commented on index.rst:
```diff
@@ -1,61 +1,190 @@
-  .. figure:: /_static/filename.ext
-     :name: fig-label
+The Butler organizes data into collections, which support read and write operations.
+Collections are organized hierarchically has paths.
```
```suggestion
Collections are organized hierarchically as paths.
```

ktlim commented on index.rst:
```diff
@@ -1,61 +1,190 @@
-     :name: fig-label
+The Butler organizes data into collections, which support read and write operations.
+Collections are organized hierarchically has paths.
+The current plan is to have the Butler service support two main path structures: one with a top-level directory per user, and one for collaborations where the top-level directory corresponds to a group.
```
```suggestion
The current plan is to have the Butler service support two user-writable path structures in addition to common data products: one with a top-level directory per user, and one for collaborations where the top-level directory corresponds to a group.
```

ktlim commented on index.rst:
```diff
@@ -1,61 +1,190 @@
+   }
+
+The Butler service may (and generally should) cache this information for a given token.
+The cache should have a maximum lifetime of an hour.
```
Maybe explain why 1 hour is appropriate and the tradeoffs of longer or shorter?

rra replied:
I reconciled this with LSE-279 (making it 30 minutes) and added a reference.

ktlim commented on index.rst:
```diff
@@ -1,61 +1,190 @@
+
+This proposal assumes that paths that do not begin with ``/u`` or ``/g`` will be written outside of the Butler service (via direct access to the Butler registry, for example) and will otherwise be public and read-only.
+
+Each collection may optionally be associated with an :abbr:`ACL (Access Control List)`.
```
```suggestion
Each collection may optionally be associated in the Butler registry with an :abbr:`ACL (Access Control List)`.
```

ktlim commented on index.rst:
```diff
@@ -1,61 +1,190 @@
+2. If the path starts with ``/u/{username}``, access will be granted and the collection created if the ``username`` attribute of the supplied token matches ``{username}``.
+3. If the path starts with ``/g/{group}``, access will be granted if ``{group}`` appears as the ``name`` field of a group in the ``groups`` attribute of the supplied token.
+
+If neither of these rules grants access, the Butler will then check the ACL again:
```
```suggestion
If neither of these rules grants access, the Butler will then check the ACL:
```

ktlim commented on index.rst:
```diff
@@ -1,61 +1,190 @@
+The identity management system will guarantee that every user is also the sole member of a group whose name matches the username.
+(This is desirable anyway for POSIX file system semantics for the Notebook Aspect of the Rubin Science Platform.)
+Therefore, to grant a specific user access to a collection, the username of the user can be added to the ACL alongside any other group.
+This satisfies the DMS-REQ-0340 requirement.
```
```suggestion
In addition, every user is a member of a single group representing all authorized users.
A collection can then be made public by adding this group to the ACL.
This satisfies the DMS-REQ-0340 requirement.
```

rra committed:
Fix typo in collection description

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

rra committed:
Fix wording for user-writable path structures

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

rra committed:
Clarify mention of ACLs

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

rra committed:
Remove stray "again" in ACL checking

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

rra committed:
Note that there is a group of all users

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

rra committed:
Clarifications based on code review

- Clarify the data products
- Say explicitly that Butler actions are via collections
- Reduce cache time to 30 minutes and add an LSE-279 reference

TallJimbo commented on index.rst:
```diff
@@ -1,61 +1,201 @@
 
-     Caption text.
+The Butler organizes data into collections, which support read and write operations.
+All Butler operations are done on collections.
```
I think we'll need to expand that; _most_ operations are, and all files and file operations are, but some of the entities that provide organizational structure for datasets (dataset types and dimension rows) are not associated with collections.

The same ownership model we use for collections will probably work for dataset types as well - those are not currently namespaced the way collections are in our internal usage, but they clearly need to be in this context.  So I think there's nothing qualitatively new about that extension.

Dimension rows will be a bit trickier, as much of our data model really depends on them being a shared resource.  They're also an "advanced usage" concept, in that most users will not ever need to define their own dimensions, and we can probably defer providing support for user-created dimensions for a while.  The main high-level use case that is difficult to support without user-owned (or at least user-inserted) dimensions is user ingest of raw data from other instruments in a way that would allow them to be processed via our middleware.  Happy to expound on the problem at some point, but a comment is probably not the most efficient method.

TallJimbo replied:
Actually, dataset types may be even simpler than that, because they can probably always be world-readable, so we may just be able to use the user/group namespacing rules with no ACLs.

rra replied:
We only care about data products from a DMS-REQ-0340 perspective, so my inclination is to change this sentence to instead say "All Butler data operations are done on collections." It doesn't sound like there's much security concern about access to dataset types or dimension rows (correct me if I'm wrong), so I think you can handle them however you want.

If you want the ACL policy for them to be recorded in this document as well so that everything's in the same place, could you propose what you want to say? If I'm understanding the problem correctly, I think saying whatever you would prefer to say is fine.

TallJimbo replied:
Sounds good.  Might be best for @timj and I to just create a new ticket or two, for one of us to add those details to this document once we've thought them through, and ask you to make sure nothing raises any flags.

rra replied:
Okay, I'll change this to data operations for now, and then am happy to review those additions.

TallJimbo reviewed (commented):
One more question I'd like to answer here (or in a related document): how do we handle dangling references?

It is possible within butler to create both collections that reference other existing collections ("CHAINED" collections; think of these as directories full of symlinks to other directories), and collections that directly reference existing datasets ("TAGGED" collections; think of these as directories full of symlinks to files).  Because those are all implemented in the database, it's so far been natural to enforce those relationships with foreign keys, and that saves us from a lot of checking for dangling references in our own code - the database guarantees that they are impossible.

I don't think it's viable to prevent users from deleting their own datasets or collections because someone they granted read access to still holds a reference to them.  It might be viable if "reference/prevent-delete" access was explicitly defined to be something more than just read access.

We could also just remove those references when the thing they refer to is deleted, but this would be highly surprising to an owner of such reference if there's no notification of it, and the easy way to implement this is with foreign keys that use `ON DELETE CASCADE`, and that gives us no way of noticing when such a delete occurred in order to provide a notification.

I think the only other option is to give up on maintaining referential integrity (for these database tables, at least) in the database, and instead add checking for dangling references in our own code.

TallJimbo commented:
One more thought on reference-based collections: they create complications for read permissions as well, in that they provide a way to read datasets via a collection other than the one intrinsic to the datasets.  I can see a few possibilities:

 - If "reference" permission is an explicit step above read permissions, we could define that it allows the holder to grant read access to datasets to other users, via those references.
 - We could define the read access lists for these reference-based collections entirely through the read access lists of their constituent datasets and collections.  In some cases we would be able to do that up-front, before generating queries; in others I'm pretty sure we'd need to check query results on the server before returning them to the client.

There may be other possibilities I haven't considered, and (like with the write-access complications raised by references), I'd like a chance to think about what the implementation constraints from the butler end might be before settling on a design.  But if we can rule out any of the possibilities for other reasons, that will simplify that investigation.

rra commented:
Do reference-based collections allow writes, or are they read-only? And are reference-based collections continuously updated if the original collection is updated, or do they only capture a point in time?

If they're read-only of a point in time, I don't think they pose any interesting security questions, so you can deal with them however you want to deal with them. I think the worst that could happen is that someone with read access to a collection would create a reference-based collection off of it and continue to have read access to that reference-based collection even if they were removed from the read ACLs of the original collection. While ideally we would avoid that, in practice this isn't very interesting from a security standpoint since someone with read access could just download the contents of the collection. There's no real way to revoke read access retroactively.

If they allow read-only access on an ongoing basis to new data put into that collection, or if they allow write access, that's more interesting, and it would be good to find a way to cut that off. Let me know if that's the case and I can brainstorm some possible ideas.

TallJimbo commented:
The reference-based collections are read-only (in that they do not allow modifications of the datasets they point to).  The CHAINED collections-of-collections variant is automatically/implicitly updated whenever the contents of one of the nested collections changes, because we evaluate those contents lazily query-by-query (we just query each nested collection in turn).  But that's a case where it'd be very easy to add access checking to the query generation logic, so it sounds like we should again just think a bit more about what we want, create a ticket for this document with that proposal, and make sure it sounds reasonable to you via review.

rra commented:
Yup, that sounds good. It might be as simple as saying that whenever a read operation is performed through a CHAINED collection, the ACL evaluated is the one on the final real collection that holds the data, not on the intermediate collection.

rra committed:
Clarify this document is on data operations

Don't claim that all Butler operations are through collections,
just data operations.

timj commented on index.rst:
```diff
@@ -1,61 +1,201 @@
+Advantages:
 
-.. .. rubric:: References
+- Simplifies group semenatics.
```
Typo: semenatics

timj reviewed (approved):
Looks okay. I'm interested in the outcome of the discussion with @TallJimbo.

rra committed:
Fix typo

rra commented:
I'm merging and publishing this, but am happy to add further revisions or review PRs as the Butler discussions continue or if anyone sees anything else.
